apps/web/components/dashboard-components/project-details/custom-llm-parameters-editor.tsx

```diff
@@ -0,0 +1,598 @@
+        customLlmParameters = {
+          ...existingParams,
+          [currentProvider]: otherModels,
+        };
+      } else {
+        // No models left for this provider - remove the provider entirely
+        const { [currentProvider]: _, ...otherProviders } = existingParams;
+        customLlmParameters =
+          Object.keys(otherProviders).length > 0 ? otherProviders : null;
+      }
+    } else {
+      // No changes needed - preserve existing parameters
+      customLlmParameters =
+        Object.keys(existingParams).length > 0 ? existingParams : null;
+    }
+
+    updateProject.mutate({
+      projectId: project.id,
+      customLlmParameters,
+    });
+  }, [project, currentProvider, currentModel, parameters, updateProject]);
+
+  const handleCancel = useCallback(() => {
+    const params = extractParameters(
+      project?.customLlmParameters,
+      currentProvider,
+      currentModel,
+    );
+    setParameters(params);
+    setIsEditing(false);
+    setHasUnsavedChanges(false);
+  }, [
+    project?.customLlmParameters,
+    currentProvider,
+    currentModel,
+    extractParameters,
+  ]);
+
+  const addParameter = useCallback(() => {
+    setParameters((prev) => [...prev, { key: "", value: "", type: "string" }]);
+    setHasUnsavedChanges(true);
+  }, []);
+
+  const removeParameter = useCallback((index: number) => {
+    setParameters((prev) => prev.filter((_, i) => i !== index));
+    setHasUnsavedChanges(true);
+  }, []);
+
+  const updateParameter = useCallback(
+    (index: number, field: keyof ParameterEntry, value: string) => {
+      setParameters((prev) => {
+        const updated = [...prev];
+        updated[index] = { ...updated[index], [field]: value };
+        return updated;
+      });
+      setHasUnsavedChanges(true);
+    },
+    [],
+  );
+
+  const applySuggestion = useCallback(
+    (suggestion: { key: string; type: string }) => {
+      if (parameters.some((p) => p.key === suggestion.key)) {
+        toast({ description: `Parameter "${suggestion.key}" already exists` });
+        return;
+      }
+
+      const defaultValue =
+        suggestion.type === "boolean"
+          ? "false"
+          : suggestion.type === "number"
+            ? "0"
+            : "";
+
+      setParameters((prev) => [
+        ...prev,
+        {
+          key: suggestion.key,
+          value: defaultValue,
+          type: suggestion.type as ParameterType,
+        },
+      ]);
+      setHasUnsavedChanges(true);
+    },
+    [parameters, toast],
+  );
+
+  // Loading state
+  if (!project) {
+    return (
+      <div className="animate-pulse space-y-4">
+        <div className="h-5 bg-muted rounded w-40" />
+        <div className="space-y-2">
+          <div className="h-4 bg-muted rounded w-full" />
+          <div className="h-4 bg-muted rounded w-3/4" />
+        </div>
+      </div>
+    );
+  }
+
+  return (
+    <div className="relative">
+      <AnimatePresence mode="wait">
+        {isEditing ? (
```

Comment:
this component is still really huge, can you break it up a bit? there's like 3 levels of ternaries, etc

I think you'll also find that having each row be its own component will make state management a little cleaner... look at headers-editor